scarb/src/ops/expand.rs

```diff
@@ -0,0 +1,214 @@
+use crate::compiler::db::{build_scarb_root_database, ScarbDatabase};
+use crate::compiler::helpers::{build_compiler_config, write_string};
+use crate::compiler::{CairoCompilationUnit, CompilationUnit, CompilationUnitAttributes};
+use crate::core::{Package, TargetKind, Workspace};
+use crate::ops;
+use crate::ops::FeaturesOpts;
+use anyhow::{anyhow, bail, Context, Result};
+use cairo_lang_compiler::db::RootDatabase;
+use cairo_lang_compiler::diagnostics::DiagnosticsError;
+use cairo_lang_defs::db::DefsGroup;
+use cairo_lang_defs::ids::{LanguageElementId, ModuleId, ModuleItemId};
+use cairo_lang_defs::patcher::PatchBuilder;
+use cairo_lang_diagnostics::ToOption;
+use cairo_lang_filesystem::db::FilesGroup;
+use cairo_lang_filesystem::ids::CrateLongId;
+use cairo_lang_formatter::{CairoFormatter, FormatOutcome, FormatterConfig};
+use cairo_lang_parser::db::ParserGroup;
+use cairo_lang_syntax::node::helpers::UsePathEx;
+use cairo_lang_syntax::node::{ast, TypedStablePtr, TypedSyntaxNode};
+use cairo_lang_utils::Upcast;
+use std::collections::HashSet;
+
+#[derive(Debug)]
+pub struct ExpandOpts {
+    pub features: FeaturesOpts,
+    pub ugly: bool,
+}
+
+pub fn expand(package: Package, opts: ExpandOpts, ws: &Workspace<'_>) -> Result<()> {
+    let package_name = package.id.name.to_string();
+    let resolve = ops::resolve_workspace(ws)?;
+    let compilation_units = ops::generate_compilation_units(&resolve, &opts.features, ws)?;
+
+    // Compile procedural macros.
+    compilation_units
+        .iter()
+        .filter(|unit| matches!(unit, CompilationUnit::ProcMacro(_)))
+        .map(|unit| ops::compile::compile_unit(unit.clone(), ws))
+        .collect::<Result<Vec<_>>>()?;
+
+    let Some(compilation_unit) = compilation_units.into_iter().find(|unit| {
+        unit.main_package_id() == package.id
+            && unit.main_component().target_kind() == TargetKind::LIB
+    }) else {
+        bail!("compilation unit not found for `{package_name}`")
+    };
+    let CompilationUnit::Cairo(compilation_unit) = compilation_unit else {
+        bail!("only cairo compilation units can be expanded")
+    };
+    let ScarbDatabase { db, .. } = build_scarb_root_database(&compilation_unit, ws)?;
+    let mut compiler_config = build_compiler_config(&compilation_unit, ws);
+    compiler_config
+        .diagnostics_reporter
+        .ensure(&db)
+        .map_err(|err| err.into())
+        .map_err(|err| {
+            if !suppress_error(&err) {
+                ws.config().ui().anyhow(&err);
+            }
+
+            anyhow!("could not check `{package_name}` due to previous error")
+        })?;
```

Comment:
LS will need to expand macros on potentially erroring projects

Reply:
I will change it in subsequent PR.